### DISCUSSION TITLE
Error by generating thumbnails

### DISCUSSION THREAD

**Discussion by rumen-vasilev** [ptarmiganlabs/butler-sheet-icons#237] [Q&A] [answered]
Hello community,

I am trying to generate some thumbnails, but I am getting an error and the process does not finish. I am using the standalone file (v2.1.3) on a Windows server machine. Does somebody know what the error on the screenshot mean and what I am doing wrong?

![image](https://user-images.githubusercontent.com/83919259/191933346-ba9ad98e-5e13-4b91-82a4-1066f525598b.png)

BR,
Rumen

**Comment by mountaindude** [comment]
Hey, I would suggest adding `--headless false` to your command line options.
That will show the underlying (normally hidden) browser that's used to connect to the Sense app and take screen shots of its sheets.
By showing that browser it's usually much easier to see where BSI gets stuck.

The error you are getting indicate that BSI (=Butler Sheet Icons) does not find the expected HTML tags/structure.

Now, this can happen in two cases: 
1. You are connecting to a virtual proxy that pops up a login screen, you have a SSO login solution or similar. This won't work as BSI expects plain form based authentication. 
2. In theory Qlik can change the code of the Sense client in new versions. Whenever Qlik releases a new version of Sense (client-managed or cloud) they can in theory change the html code, breaking BSI. As of this writing BSI has been tested with 2022 May version of client-managed Qlik Sense.

**Reply by rumen-vasilev** [reply]
Hello Göran,

I am using actually SSO login. It's a little bit, that I can not use tlis tool on our site. Thank you very much for the explaination.

BR,
Rumen

**Comment by mountaindude** [comment, marked as answer]
That explains it then. You are most likely getting some login popup.

Is it possible to create a separate virtual proxy in Sense that is only
used for API access from devops or admin tools?
That's how many Sense admins solve this challenge.
You can also lock down that virtual proxy with firewall rules etc, making
it available only from select IP ranges or computers.

**Reply by rumen-vasilev** [reply]
Yes, I will try to create a separate virtual proxy. Should I use the SAML authentification method for this purpose?
![image](https://user-images.githubusercontent.com/83919259/194335018-e9aff2da-2b38-420e-8753-46dbfee8acd5.png)

**Comment by mountaindude** [comment]
Long overdue response, sorry for that.

No, SAML would not be my first choice here. 
Sure, you *could* use SAML but you would then have to set up all the SAML configuration - which can be a challenge.

I would go with a regular ticket based virtual proxy, but it's **important** that it uses `Form` authentication. 
I.e. the username and password should be entered on a web page. 
This is needed because Butler Sheet Icons pretends to be a regular user when taking the screen shots of the Sense app sheets - and also when logging into Sense. A Windows login popup won't work then.

My virtual proxy config looks like this:

<img width="1097" alt="image" src="https://user-images.githubusercontent.com/1029262/210444540-10ca893f-3a02-453d-8cc2-e52c5e099d22.png">

<img width="1094" alt="image" src="https://user-images.githubusercontent.com/1029262/210444675-526e152f-223c-4ee8-bd1d-672193730990.png">

Finally, using Butler Sheet Icons version 3.0 (which was released today!) I run this command (using PowerShell on Windows Server 2016 in this case). Note the final `--prefix form`, which tells Butler Sheet Icons to use a virtual proxy called `form`.

`PS C:\tools\butler-sheet-icons-win\demo-dir> .\butler-sheet-icons.exe qseow create-sheet-thumbnails --host $BSI_HOST --appid $BSI_APP_ID --apiuserdir Internal --apiuserid sa_
api --loglevel info --logonuserdir $BSI_LOGON_USER_DIR --logonuserid $BSI_LOGON_USER_ID --logonpwd $BSI_LOGON_PWD --contentlibrary 'abc 123' --pagewait 5 --secure true --head
less true --imagedir './img' --certfile $BSI_CERT_FILE --certkeyfile $BSI_CERT_KEY_FILE --includesheetpart 4 --sense-version $BSI_SENSE_VERSION --prefix form`

**Reply by rumen-vasilev** [reply]
Hello Göran,

Thanks for your suggestion. I will try it.

BR,
Rumen
